executor/cook/executor.py

```diff
@@ -0,0 +1,364 @@
+        process = subprocess.Popen(command, shell=True, stdout=stdout, stderr=stderr)
+
+        return process, stdout, stderr
+    except Exception:
+        logging.exception('Error in launch_task')
+        update_status(driver, task, 'TASK_FAILED')
+
+
+def is_running(process):
+    """
+        :param process: The process to query
+        :return: whether the process is still running.
+    """
+    return process.poll() is None
+
+
+def has_crashed(process):
+    """
+        :param process:
+        :return:
+    """
+    return not is_running(process) and process.poll() is not 0
+
+
+def cleanup_process(process_info):
+    """
+        :param process_info: Tuple of (process, stdout_file, stderr_file)
+        :return: Nothing
+    """
+    _, stdout, stderr = process_info
+    logging.info('Closing {}'.format(stdout.name))
+    stdout.close()
+    logging.info('Closing {}'.format(stderr.name))
+    stderr.close()
+
+
+def kill_task(driver, task, process_info):
+    """
+        :param driver: The mesos driver to use.
+        :param task: The task to execute.
+        :param process_info: Tuple of (process, stdout_file, stderr_file)
+        :return: Nothing
+    """
+    try:
+        process, _, _ = process_info
+        if is_running(process):
+            logging.info('Killing task')
+            process.kill()
+        cleanup_process(process_info)
+
+        update_status(driver, task, 'TASK_KILLED')
+    except:
+        logging.exception('Error in kill_task')
+        update_status(driver, task, 'TASK_FAILED')
+
+
+def await_process_completion(driver, task, stop_signal, process_info):
+    """
+        :param driver: The mesos driver to use.
+        :param task: The task to execute.
+        :param stop_signal: Event that determines if an interrupt was sent
+        :param process_info: Tuple of (process, stdout_file, stderr_file)
+        :return: Truw if the process was killed, False if it terminated naturally.
+    """
+    process, _, _ = process_info
+
+    process_killed = False
+    while is_running(process):
+
+        if stop_signal.isSet():
+            logging.info('Executor has been instructed to terminate')
+            kill_task(driver, task, process_info)
+            process_killed = True
+            break
+
+        time.sleep(1)
+    return process_killed
+
+
+def manage_task(driver, task, stop_signal, completed_signal, max_message_length, progress_poll_interval_ms,
+                sandbox_location, stdout_file, stderr_file):
+    """
+        Manages the execution of a task.
+        :return: Nothing
+    """
+    try:
+        task_id = get_task_id(task)
+        sandbox_message = json.dumps({'sandbox-location': sandbox_location, 'task-id': task_id})
+        send_message(driver, sandbox_message, max_message_length)
+
+        process_info = launch_task(driver, task, stdout_file, stderr_file)
+        process, _, _ = process_info
+
+        progress_watcher = ProgressWatcher(stdout_file, completed_signal)
```

Comment:
A job should be able to configure where to read progress from as it might not make sense to write to stdout always.